### PR TITLE
[onert] Add unit tests for training arithmetic ops

### PR DESCRIPTION
**Pull request by ragmani**
This commit adds some tests for arithmetic ops to validate training.
   - Introduce BranchModelTrain.test.cc that validates branched cases
   - Add tests that validates arithmetic ops(Add, Mul, Sub)

ONE-DCO-1.0-Signed-off-by: ragmani <ragmani0216@gmail.com>

### CODE REVIEW

File: tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    int32_t batch_size = 1;
+    cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                       circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
+                       circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
+
+    _context = std::make_unique<GenModelTrainContext>(cgen.finish());
+    _context->addTrainCase(uniformTCD<float>(
+      {{{1, 3}, {0, 1, 2, 3, 4, 5, 6, 7}}, {{2, 1}, {7, 6, 5, 4, 3, 2, 1, 0}}}, // inputs
+      {{{2, 1, 5, 5, 2, 1, 5, 5}}, {{2, 1, 5, 5, 2, 1, 5, 6}}},                 // expected
+      {8.4678f}                                                                 // loss
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc:50]: The result of executing the same model with the same data on tensorflow:
```
Epoch 1/5
2/2 [==============================] - 0s 3ms/step - loss: 9.2218 - categorical_accuracy: 0.0000e+00
Epoch 2/5
2/2 [==============================] - 0s 2ms/step - loss: 8.9554 - categorical_accuracy: 0.0000e+00
Epoch 3/5
2/2 [==============================] - 0s 1ms/step - loss: 8.7044 - categorical_accuracy: 0.0000e+00
Epoch 4/5
2/2 [==============================] - 0s 1ms/step - loss: 8.4678 - categorical_accuracy: 0.0000e+00
Epoch 5/5
2/2 [==============================] - 0s 1ms/step - loss: 8.2444 - categorical_accuracy: 0.0000e+00
```

File: tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    int32_t batch_size = 1;
+    cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                       circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
+                       circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
+
+    _context = std::make_unique<GenModelTrainContext>(cgen.finish());
+    _context->addTrainCase(uniformTCD<float>(
+      {{{0, 1, 2, 3, 4, 5, 1, 3}, {6, 7}}, {{5, 4, 3, 2, 1, 0, 2, 1}, {7, 6}}}, // inputs
+      {{{2, 1, 5, 5, 2, 1, 5, 5}}, {{2, 1, 5, 5, 2, 1, 5, 6}}},                 // expected
+      {3.2863f}                                                                 // loss
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc:91]:
```
Epoch 1/5
2/2 [==============================] - 0s 3ms/step - loss: 7.3265 - categorical_accuracy: 0.0000e+00
Epoch 2/5
2/2 [==============================] - 0s 1ms/step - loss: 4.6811 - categorical_accuracy: 0.0000e+00
Epoch 3/5
2/2 [==============================] - 0s 1ms/step - loss: 3.6735 - categorical_accuracy: 0.0000e+00
Epoch 4/5
2/2 [==============================] - 0s 1ms/step - loss: 3.2863 - categorical_accuracy: 0.0000e+00
Epoch 5/5
2/2 [==============================] - 0s 1ms/step - loss: 3.1329 - categorical_accuracy: 0.0000e+00
```

File: tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    int32_t batch_size = 1;
+    cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                       circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
+                       circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
+
+    _context = std::make_unique<GenModelTrainContext>(cgen.finish());
+    _context->addTrainCase(
+      uniformTCD<float>({{{0, 3}, {6, 7}}, {{5, 4}, {7, 6}}},                     // inputs
+                        {{{3, 2, 1, 2, 5, 6, 1, 0}}, {{2, 1, 5, 5, 2, 1, 5, 6}}}, // expected
+                        {12.2822f}                                                // loss
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc:144]:
```
Epoch 1/5
2/2 [==============================] - 0s 3ms/step - loss: 12.5488 - categorical_accuracy: 0.0000e+00
Epoch 2/5
2/2 [==============================] - 0s 1ms/step - loss: 12.4590 - categorical_accuracy: 0.0000e+00
Epoch 3/5
2/2 [==============================] - 0s 1ms/step - loss: 12.3701 - categorical_accuracy: 0.0000e+00
Epoch 4/5
2/2 [==============================] - 0s 1ms/step - loss: 12.2822 - categorical_accuracy: 0.0000e+00
Epoch 5/5
2/2 [==============================] - 0s 1ms/step - loss: 12.1951 - categorical_accuracy: 0.0000e+00
```

**Review by hseok-oh — APPROVED** [review]
LGTM

**Review by zetwhite — APPROVED** [review]
LGTM :+1:

**Review by Aeren1564 — APPROVED** [review]
LGTM!

File: tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    cgen.addOperatorFullyConnected({{input0, weight, bias}, {fc_output}});
+    cgen.addOperatorAdd({{fc_output, input1}, {output}},
+                        circle::ActivationFunctionType::ActivationFunctionType_NONE);
+    cgen.setInputsAndOutputs({input0, input1}, {output});
+
+    float learning_rate = 0.01f;
+    int32_t batch_size = 1;
+    cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                       circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
+                       circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc:44]: If #13199 would be merged first,
```suggestion
                       circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
    cgen.markAllOpsAsTrainable();
```

Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc:44]: If #13199 would be merged first,
```suggestion
                       circle::LossReductionType::LossReductionType_SumOverBatchSize, batch_size});
    cgen.markAllOpsAsTrainable();
```

File: tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    cgen.addOperatorFullyConnected({{input1, weight1, bias1}, {fc_output1}});
+    cgen.addOperatorMul({{fc_output0, fc_output1}, {mul_output}},
+                        circle::ActivationFunctionType::ActivationFunctionType_NONE);
+    cgen.addOperatorFullyConnected({{mul_output, weight2, bias2}, {output}});
+    cgen.setInputsAndOutputs({input0, input1}, {output});
+
+    float learning_rate = 0.01f;
+    int32_t batch_size = 1;
+    cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                       circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/BranchModelTrain.test.cc:137]:
```suggestion
                       circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
        cgen.markAllOpsAsTrainable();
```

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Add.test.cc [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  int lhs = cgen.addTensor({{1, 2, 2, 1}, circle::TensorType::TensorType_FLOAT32});
+  int rhs = cgen.addTensor({{1, 2, 3, 1}, circle::TensorType::TensorType_FLOAT32});
+  int out = cgen.addTensor({{1, 2, 3, 1}, circle::TensorType::TensorType_FLOAT32});
+  cgen.addOperatorAdd({{lhs, rhs}, {out}}, circle::ActivationFunctionType_NONE);
+  cgen.setInputsAndOutputs({lhs, rhs}, {out});
+
+  float learning_rate = 0.01f;
+  int32_t batch_size = 1;
+  cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Add.test.cc:33]:
```suggestion
                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
        cgen.markAllOpsAsTrainable();
```

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Add.test.cc [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  int lhs = cgen.addTensor({{1, 2, 2, 1}, circle::TensorType::TensorType_INT32});
+  int rhs = cgen.addTensor({{1, 2, 3, 1}, circle::TensorType::TensorType_FLOAT32});
+  int out = cgen.addTensor({{1, 2, 3, 1}, circle::TensorType::TensorType_FLOAT32});
+  cgen.addOperatorAdd({{lhs, rhs}, {out}}, circle::ActivationFunctionType_NONE);
+  cgen.setInputsAndOutputs({lhs, rhs}, {out});
+
+  float learning_rate = 0.01f;
+  int32_t batch_size = 1;
+  cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Add.test.cc:53]:
```suggestion
                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
        cgen.markAllOpsAsTrainable();
```

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mul.test.cc [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  int lhs = cgen.addTensor({{1, 2, 2, 1}, circle::TensorType::TensorType_FLOAT32});
+  int rhs = cgen.addTensor({{1, 2, 3, 1}, circle::TensorType::TensorType_FLOAT32});
+  int out = cgen.addTensor({{1, 2, 3, 1}, circle::TensorType::TensorType_FLOAT32});
+  cgen.addOperatorMul({{lhs, rhs}, {out}}, circle::ActivationFunctionType_NONE);
+  cgen.setInputsAndOutputs({lhs, rhs}, {out});
+
+  float learning_rate = 0.01f;
+  int32_t batch_size = 1;
+  cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mul.test.cc:33]:
```suggestion
                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
        cgen.markAllOpsAsTrainable();
```

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mul.test.cc [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  int lhs = cgen.addTensor({{1, 2, 2, 1}, circle::TensorType::TensorType_INT32});
+  int rhs = cgen.addTensor({{1, 2, 3, 1}, circle::TensorType::TensorType_FLOAT32});
+  int out = cgen.addTensor({{1, 2, 3, 1}, circle::TensorType::TensorType_FLOAT32});
+  cgen.addOperatorMul({{lhs, rhs}, {out}}, circle::ActivationFunctionType_NONE);
+  cgen.setInputsAndOutputs({lhs, rhs}, {out});
+
+  float learning_rate = 0.01f;
+  int32_t batch_size = 1;
+  cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Mul.test.cc:53]:
```suggestion
                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
        cgen.markAllOpsAsTrainable();
```

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Sub.test.cc [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+  int lhs = cgen.addTensor({{1, 2, 2, 1}, circle::TensorType::TensorType_FLOAT32});
+  int rhs = cgen.addTensor({{1, 2, 3, 1}, circle::TensorType::TensorType_FLOAT32});
+  int out = cgen.addTensor({{1, 2, 3, 1}, circle::TensorType::TensorType_FLOAT32});
+  cgen.addOperatorSub({{lhs, rhs}, {out}}, circle::ActivationFunctionType_NONE);
+  cgen.setInputsAndOutputs({lhs, rhs}, {out});
+
+  float learning_rate = 0.01f;
+  int32_t batch_size = 1;
+  cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Sub.test.cc:34]:
```suggestion
                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
        cgen.markAllOpsAsTrainable();
```

File: tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Sub.test.cc [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  int lhs = cgen.addTensor({{1, 2, 2, 1}, circle::TensorType::TensorType_INT32});
+  int rhs = cgen.addTensor({{1, 2, 3, 1}, circle::TensorType::TensorType_FLOAT32});
+  int out = cgen.addTensor({{1, 2, 3, 1}, circle::TensorType::TensorType_FLOAT32});
+  cgen.addOperatorSub({{lhs, rhs}, {out}}, circle::ActivationFunctionType_NONE);
+  cgen.setInputsAndOutputs({lhs, rhs}, {out});
+
+  float learning_rate = 0.01f;
+  int32_t batch_size = 1;
+  cgen.addTrainInfo({circle::Optimizer::Optimizer_SGD, learning_rate,
+                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
```
Comment (ragmani) [inline tests/nnfw_api/src/GenModelTests/nontrainable_op_trains/Sub.test.cc:54]:
```suggestion
                     circle::LossFn::LossFn_MEAN_SQUARED_ERROR,
        cgen.markAllOpsAsTrainable();
```

**Comment by hseok-oh** [comment]
@ragmani  https://github.com/Samsung/ONE/pull/13199 is merged

**Comment by ragmani** [comment]
> @ragmani https://github.com/Samsung/ONE/pull/13199 is merged@ragmani https://github.com/Samsung/ONE/pull/13199 is merged

Added it :)

**Review by hseok-oh — APPROVED** [review]

**Review by jyoungyun — APPROVED** [review]
LGTM

**Pull request merged**